pkg/controller/controller_binding.go

```diff
@@ -347,38 +350,69 @@ func isPlanBindable(serviceClass *v1alpha1.ServiceClass, plan *v1alpha1.ServiceP
 
 func (c *controller) injectBinding(binding *v1alpha1.Binding, credentials map[string]interface{}) error {
 	glog.V(5).Infof("Creating Secret %v/%v", binding.Namespace, binding.Spec.SecretName)
-	secret := &v1.Secret{
-		ObjectMeta: metav1.ObjectMeta{
-			Name:      binding.Spec.SecretName,
-			Namespace: binding.Namespace,
-		},
-		Data: make(map[string][]byte),
-	}
 
+	secretData := make(map[string][]byte)
 	for k, v := range credentials {
 		var err error
-		secret.Data[k], err = serialize(v)
+		secretData[k], err = serialize(v)
 		if err != nil {
 			return fmt.Errorf("Unable to serialize credential value %q: %v; %s",
 				k, v, err)
 		}
 	}
 
-	found := false
+	secretClient := c.kubeClient.Core().Secrets(binding.Namespace)
+	// Retry loop for upserting the Secret
+	for {
+		found := true
+		existingSecret, err := secretClient.Get(binding.Spec.SecretName, metav1.GetOptions{})
+		if err != nil {
+			if !errors.IsNotFound(err) {
+				return err
+			}
+			found = false
+		}
 
-	_, err := c.kubeClient.Core().Secrets(binding.Namespace).Get(binding.Spec.SecretName, metav1.GetOptions{})
-	if err == nil {
-		found = true
-	}
+		if found {
+			if !IsOwnedBy(existingSecret, binding) {
+				return fmt.Errorf("Secret '%s' is not owned by Binding '%s', ownerReferences: %v",
+					existingSecret.Name, binding.Name, existingSecret.OwnerReferences)
```

Comment:
Good point. I will remove mentioning Binding details completely, and let the invoking side to add the Binding namespace/name for logging, or putting the message to the Binding status (where Binding details are not needed).